Development

Tasenga opened the pull request:
changes from the PR#1 were made

Commit:
using dockstring inside body functions

Commit:
Function 'File_creation' was moved to separate module and is imported.

Commit:
Code was optimized:

- creating the set of keys and using the fromkeys method  to creating dictionary were replaced by strings in existing loops;
- creating the lists of positive and negative messages were combined into one function 'message_list'.
- creating the list of popular words for positive and negative messages were combined into one function 'popular_word_list'.

Commit:
1. Function avg_rating is fixed:

- used a list_comprehensive to keep the code more concise,
- added boolean parameter "voters", which determines to show the number of voters or not.
2. Function avg_rating is imported in Task_2.py.

minor fix: output of function "nonanalys_data" to file was corrected

Commit:
The codes of functions 'best_comment' and 'bad_comment' was optimized by using min and max methods

Commit:
Function "nearest_reviews" changed:

- function was separated into small methods,
- added docstrings for all function's methods,
- added comments for function's main arguments,
- code was optimized by using following basic methods: min, filter, sorted and also list_comprehensive.

sleshJdev commented on Task_1.py:
```diff
@@ -1,160 +1,165 @@
+        overall, count = asins_value.get(item['asin'], [0, 0])
+        asins_value[item['asin']] = [overall + item['overall'], count + 1]
+    if not voters:
+        avg_overall = [[key, str(round(value[0] / value[1], 2)).replace('.', ',')]
```
Why we make this manipulations with converting float -> string and replace dots with commas?

Tasenga replied:
It need that to check results. Because some viewers, like excel, change int with comma on date.
Do I need to delete this manipulations?

sleshJdev replied:
Let's keep, just do it in more ideomatic way using . format function

sleshJdev commented on Task_1.py:
```diff
@@ -1,160 +1,165 @@
+        for key, value in main_dict.items():
+            if 'reviewerName' in value.keys() and 'unixReviewTime' in value.keys() \
+                    and value['reviewerName'] != 'Amazon Customer':
+                try:
```
Could you please clarify the purpose of try-except here? If we deal with incorrect data in such a way then it's not idiomatic. It's desired to split filtering and processing phases. Let's prepare data, and then process it.

Tasenga replied:
Try is used when the key has already existed in the dictionary.
Except - to create a new dictionary item.
I can split filtering and processing phases, but I thing that try-except will stay. Do it?

sleshJdev replied:
No, we defenetly don't need try-except here

sleshJdev commented on Task_1.py:
```diff
@@ -1,160 +1,165 @@
+                    all_timereview.setdefault(value['reviewerName'], {key: value['unixReviewTime']})
+        return all_timereview
+
+    def deltatime(all_timereview=dict_timereviews_of_user(main_dict)):
```
Please, call _deltatime_ with desired params explicitly. In this particular case default value is ill-timed as it's too specific.

Tasenga replied:
I hope I right understand it.
`def deltatime(all_timereview)):`
And then, when deltatime will called I need to transfer all_timereview=dict_timereviews_of_user(main_dict) to function. Right?

sleshJdev replied:
No, I meant that it's not obvious when you declare the default parameter without actual need.
So, I'd that def _deltatime_ has signature like `def deltatime(all_timereview)`, then when calling it you explicitly pass required parameters:
```
deltas = deltatime(dict_timereviews_of_user(main_dict))
```

sleshJdev commented on Task_1.py:
```diff
@@ -1,160 +1,165 @@
-                a = i-1
-                deltatime.append([sort[i][0], sort[a][0], sort[i][1] - sort[a][1]])
-    time_deltatime = []
+def nearest_reviews(main_dict):
```
Could you please move functions declared inside _nearest_reviews_ move to outside. It will make _nearest_reviews_ cleaner.

Tasenga replied:
Will do!

sleshJdev commented on Task_1.py:
```diff
@@ -1,160 +1,165 @@
+                                                ...}
+                        ...}'''
+        all_timereview = {}
+        for key, value in main_dict.items():
```
let's try to implement it using use for-comrehensive + groupby + sorted. I'd expect something like:
```
unix_sorted_reviews_by_names = {
  name: sorted(reviews, key = lambda review: review['unixReviewTime'])
  for name, reviews 
  in groupby(rows, key = lambda review: review['reviewerName'])
}
unix_max_diff_per_name = {
  name: reviews[-1]['unixReviewTime'] - reviews[0]['unixReviewTime']
  for name,reviews 
  in unix_sorted_reviews_by_names.items() 
  if len(reviews) > 1
}
```

Tasenga replied:
I will try to add 'groupby'.

sleshJdev commented on Task_1.py:
```diff
@@ -1,160 +1,165 @@
+        return deltatime
+
+    nearest_reviews = min(
+                    list(filter(lambda comment: comment[2] != 0, deltatime())),
```
`list(filter(lambda comment: comment[2] != 0, deltatime()))` called twice, it's not very efficient

sleshJdev commented on Task_1.py:
```diff
@@ -1,160 +1,165 @@
+                    key=lambda comment: comment[2])
+
+    # argument 'analyzed_data' shows how much reviews were included in 'nearest_review' list
     global analyzed_data
```
Please, let's get rid of global. It's not functional. Just pass it as parameter

sleshJdev commented on Task_1.py:
```diff
@@ -1,160 +1,165 @@
+    the application which received the most useless message;
+    '''
+    bad_comment = min(
+        dict(filter(lambda review: review[1]['helpful'][1] != 0, main_dict.items())).values(),
```
We create `dict`, then call `.values()`, maybe just do not create it?

sleshJdev commented on Task_2.py:
```diff
@@ -1,34 +1,12 @@
-        writer = csv.writer(csv_file, delimiter='\t')
-        for line in data:
-            writer.writerow(line)
+from Task_1 import avg_rating
```
avg_rating should be in separate module, but not in Task_X file. Because, Task_X files should know  nothing about each other. Just craete a dedicated module e.g. _statistics_ and put there all re-usable functions about calculating metrics etc.

sleshJdev commented on Task_3.py:
```diff
@@ -1,56 +1,38 @@
-    for word in word_positive_comment:
-        popular_in_positive_comment[word] = popular_in_positive_comment.get(word) + 1
-    return popular_in_positive_comment
+            if item['overall'] >= 2.5:
```
What about this:
```
words = re.split(r'\W+', item['reviewText'].replace("'", '_').lower())
(positive_comments if item['overall'] >= 2.5 else negative_comments).extend(words)
```
?

Tasenga replied:
I found only such variant.
```
def get_text(main_dict, overall_values):
    '''function gets a text and returns list of reviews in depends on overall values'''
    message_list = [item['reviewText'] for item in main_dict.values()
                    if 'overall' in item.keys() and item['overall'] in overall_values]
    return message_list

def split_text_up(text):
    '''function gets a text and returns list of words from this text'''
    list_of_words = []
    for review in text:
        list_of_words.extend(re.split(r'\W+', review.replace("'", '_').lower()))
    return list_of_words
```

Tasenga replied:
I worry about this part:
```
   list_of_words = []
    for review in text:
        list_of_words.extend(re.split(r'\W+', review.replace("'", '_').lower()))
```
but I really don't know how do it simpler, cause I need result like ['w1', 'w2',...., 'wn']
and when I write such code 
`list_of_words = [word for word in [re.split(r'\W+', review.replace("'", '_').lower()) for review in text]]`
I get the result like [ ['w1', 'w2', ....] , [...], ....., ['wn', 'wm', ...] ]

sleshJdev commented on document_creation.py:
```diff
@@ -0,0 +1,21 @@
+import csv
+
+
+class file_create:
```
In this case, file_create doesn't have state(no fields, not data) so we don't need to have it as a class. Let's rather declare pure functions and import them in place.

sleshJdev replied:
Let's name it create_directory? Usually function names are verbs.

Tasenga replied:
Will do!

sleshJdev commented on Task_1.py:
```diff
@@ -1,160 +1,165 @@
-
+    file_name = 'general-stats.cvs'
+    file_create().save_file(file_name, avg_rating(main_dict))
+    file_create().save_file(file_name, best_comment(main_dict), 'a')
```
Import save_file func directly, not need to create uselessly object each time.

Tasenga replied:
Will do!

sleshJdev commented on Task_1.py:
```diff
@@ -1,160 +1,165 @@
+    comment.append([count_unanalyzed_2, ' - to get the application which received the most useless message'])
+
+    count_unanalyzed_3 = len(main_dict) - analyzed_data
+    comment.append([str(count_unanalyzed_3) + ' or ' + str(round(count_unanalyzed_3/len(main_dict) * 100)) + '% ',
```
Please, use [string interpolation](https://www.python.org/dev/peps/pep-0498/), it's much more convenient

Tasenga replied:
I see. Will do!

sleshJdev commented on Task_1.py:
```diff
@@ -1,160 +1,165 @@
-        avg_overall.append([key, asins_value.get(key)[0] / asins_value.get(key)[1]])
+        overall, count = asins_value.get(item['asin'], [0, 0])
+        asins_value[item['asin']] = [overall + item['overall'], count + 1]
+    if not voters:
```
Do we really need **voters** parameter? Let's rather always return likes too and the caller will decide how to handle them.

Tasenga replied:
There was such requirements in individual_task:
''2. На основании поданного на вход файла, сгенерировать файл general-stats.cvs содержащий:    a) Среднюю оценка каждого приложения ''
''3. Сгенерировать файл apps-stats.cvs с информацией о средней оценке каждого приложения и количестве проголосовавших пользователей''

All the same to remove the number of voters?

sleshJdev replied:
I meant that function returned value structure depends on the flag that is error prone as caller will have to check this flag as well to know how to handle function result. Instead, let's always return two values tuple

Commit:
class 'file_create' was splited up on 2 functions 'create_directory',  'save_file'.

Commit:
Function 'message_list' was separated for 2 functions 'get_text' and 'split_text_up' to optimize this part of code:

>>>   for item in main_dict.values():
>>>        	if 'overall' in item.keys():
>>>           		if item['overall'] >= 2.5:
>>>                			positive_comments.extend(re.split(r'\W+', item['reviewText'].replace("'", '_').lower()))
>>>           		else:
>>>                 			negative_comments.extend(re.split(r'\W+', item['reviewText'].replace("'", '_').lower()))

Commit:
small fix related with changing function 'save_file'

Commit:
More ideomatic way ({}.format) was used to get the arguments in functions 'avg_raiting' , 'best_comment', 'bad_comment', 'nonanalys_data';

Function 'nearest_reviews' has been completely changed with such methods like filter, groupby etc.;
Functions 'deltatime' and 'all_timereview' have been removed because don't use;
Parameter 'voters' has been canceled from function 'avg_raiting'. Quantities of output values is determined in output process.

Commit:
Added information about configuration local git hooks to README.

Commit:
Files moved in new directory because new module was created.

Function "nearest_review" from Task_1_1 was splited in small functions:
- 'sort_dict_by_name' returns the sorted dictionary, where keys are ReviewerID and values are list of all user reviews per user
- 'create_dict_diff_unixtime' returns the dictionary, where keys are two ReviewText (delimiter = '[delimiter]') and values are difference between unixReviewTime for 'reviews' in key and reviewerID who sent this reviewText
- 'find_potential_bots' - returns the set of potential bots. It is assumed that a potential bot may be a user who leaves several reviews per second
- 'nearest_reviews' - returns string with information about the shortest interval between reviews of one user and the length of both messages

Commit:
format  'replace('.', ',')' for  'round(value[0] / value[1], 2)' was deleted;

function 'nearest_reviews' was changed: it is consist of two methods:
- 'count_deltatime' -  function returns information per user about the two nearest reviews were sent by this user;
- 'nearest_reviews' - function returns the shortest interval between reviews of one user and the length of both messages.
Also parameter 'number_unanalyzed_reviews' was added which counts how much reviews was excluded from the analyze

Commit:
unused files were deleted

Commit:
git config --local core.hooksPath git.hooks

Commit:
removed old files

Commit:
deleted info about configuration git.hooks

Commit:
correct BASE_DIR to right work 'os.walk'

Tasenga committed:
Merge pull request #5 from Tasenga/L2_git_hooks

L2 git hooks

Commit:
rewrite getting main_dict

Commit:
Merge branch 'development' of https://github.com/Tasenga/ITechArt_Android_App_Review into development
